### PR TITLE
fix virtctl create vm volume-import flag

### PR DESCRIPTION
**Pull request by opokornyy**
**What this PR does / why we need it**:

Using the `--volume-import` flag with `PVC` and `Snapshot` types may result in VMs taking ownership of cloned DVs due to the incorrect use of the name parameter. To resolve this issue, we are introducing the src parameter for the `--volume-import` flag.

**Which issue(s) this PR fixes**
Fixes https://github.com/kubevirt/kubevirt/issues/10592

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
```
None
```

### CODE REVIEW

**Comment by alicefr** [comment]
@0xFelix @opokornyy does the new formatting break any old behavior? AFAIU, we don't gurantee any backward compatibility between `virtctl` versions, do you have the same understanding?

File: pkg/virtctl/create/vm/vm.go [hunk 1]
```diff
@@ -1206,14 +1215,23 @@ func withVolumeSourceSnapshot(paramStr string, vm *v1.VirtualMachine) (*cdiv1.Da
 		return nil, err
 	}
 
-	if sourceStruct.Name == "" || sourceStruct.Namespace == "" {
-		return nil, params.FlagErr(VolumeImportFlag, "name and namespace are both required with Snapshot volume source")
+	if sourceStruct.Source == "" {
+		return nil, params.FlagErr(VolumeImportFlag, "src must be specified")
+	}
+
+	namespace, name, err := params.SplitPrefixedName(sourceStruct.Source)
```
Comment (alicefr) [inline pkg/virtctl/create/vm/vm.go:1222]: What if the user only specifies the pvc claim and not the namespace, doesn't the claim become the `namespace` and `name` is empty?

Comment (0xFelix) [reply pkg/virtctl/create/vm/vm.go:1222]: Right, good catch. The removed check in L1209 needs to be moved to L1227.

File: pkg/virtctl/create/vm/vm.go [hunk 2]
```diff
@@ -1104,14 +1104,23 @@ func withVolumeSourcePVC(paramStr string, vm *v1.VirtualMachine) (*cdiv1.DataVol
 		return nil, err
 	}
 
-	if sourceStruct.Name == "" || sourceStruct.Namespace == "" {
```
Comment (0xFelix) [inline pkg/virtctl/create/vm/vm.go:1107]: Same as below, we need this check in L1116.

File: pkg/virtctl/create/vm/vm_test.go [hunk 3]
```diff
@@ -922,7 +922,14 @@ var _ = Describe("create vm", func() {
...
+			Entry("Invalid src in pvc volume source", "failed to parse \"--volume-import\" flag: src must be specified", setFlag(VolumeImportFlag, "type:pvc,size:256Mi,src:")),
+			Entry("Missing src namespace in pvc volume source", "failed to parse \"--volume-import\" flag: namespace of pvc 'my-pvc' must be specified", setFlag(VolumeImportFlag, "type:pvc,size:256Mi,src:/my-pvc")),
+			Entry("Missing src name in pvc volume source", "failed to parse \"--volume-import\" flag: src invalid: name cannot be empty", setFlag(VolumeImportFlag, "type:pvc,size:256Mi,src:default/")),
+			Entry("Missing src in snapshot volume source", "failed to parse \"--volume-import\" flag: src must be specified", setFlag(VolumeImportFlag, "type:pvc,size:256Mi")),
+			Entry("Invalid src in snapshot volume source", "failed to parse \"--volume-import\" flag: src must be specified", setFlag(VolumeImportFlag, "type:snapshot,size:256Mi")),
+			Entry("Missing src namespace in snapshot volume source", "failed to parse \"--volume-import\" flag: namespace of snapshot 'my-snapshot' must be specified", setFlag(VolumeImportFlag, "type:snapshot,size:256Mi,src:/my-snapshot")),
```
Comment (0xFelix) [inline pkg/virtctl/create/vm/vm_test.go:931]: Can you please add a test case for the issue found above? e.g. `src:noslashsinglevalue`

**Comment by 0xFelix** [comment]
@alicefr Right, this breaks the old behavior, but I have to agree, there is no backwards compatibility guaranteed. IMO this way is less error prone (see #10592) and in line with how other flags of this cmd work. (`src` to specify source pvc, `name` to optionall provide a name for the volume`)

File: pkg/virtctl/create/vm/vm.go (hunk 1 above)
Comment (0xFelix) [reply pkg/virtctl/create/vm/vm.go:1222]: Actually, `params.SplitPrefixedName` has named return values and will return empty namespace + name for e.g. `stringwithoutslash`. It also ensures `name` is not empty and the the following check will ensure `namespace` is not empty. So this should be enough.

Comment (opokornyy) [reply pkg/virtctl/create/vm/vm.go:1222]: It works exactly as Felix said, the name is checked in `SplitPrefixedName`-  https://github.com/kubevirt/kubevirt/blob/d7371feb56fcb138806b1b6e5b6ab154e900a2e0/pkg/virtctl/create/params/params.go#L170-L185 There is also one test case for this situation here - https://github.com/kubevirt/kubevirt/pull/10627/files#diff-57e5aa5a774e43ac18df44560d637ce9003cfe9f8d92fd9eaaec4d4302416832R928

**Review by 0xFelix — APPROVED** [review]
Thanks!

/lgtm
/retest-required

**Comment by alicefr** [comment]
The code looks good, thanks for the work!
/approve

**Review by 0xFelix — APPROVED** [review]
/lgtm

Thanks!

**Pull request merged**